(BSR) refactor(sonarqube): Sonarqube security issues (random seeds in sklearn functions and security check for curl in dockerfiles)

cibrahim-pass opened the pull request:
# PR template

## Select PR template in preview mode

* [Hotfix](?expand=1&template=Hotfix_template.md)
* [DA](?expand=1&template=DA_ticket_template.md)
* [DE](?expand=1&template=DE_ticket_template.md)
* [DS](?expand=1&template=DS_ticket_template.md)
* [MEP](?expand=1&template=MEP_template.md)

## PR title format (except for MEP)

There is a linter on the PR title format. Please respect the following format:

<details>
<summary>(ticket) type(topic): comment</summary>

* ticket surrounded by parenthesis, with optionnaly a hyphen followed by one or more digits (e.g., -1234). The first part must be one of the following strings:
  * DA
  * DE
  * AE
  * DS
  * HF
  * BSR
  * PC

* type :
The second part to specify the type of change one of the following :
  * build
  * lint
  * ci
  * docs
  * feat
  * fix
  * perf
  * refactor
  * test
  * core
  * dbt

* topic within parenthesis: 1 word e.g., (dag)

* comment: tell us your life

examples:

* :white_check_mark: (DE-124) refactor(firebase): update source field
* :x: (DE-124) refactor (firebase): update source field **(space between type and topic)**
* :x: (DE-124) airflow(firebase): update source fiedd in DAG **(wrong type)**
* :x: (DE-124) (DE-124) refactor(firebase refacto): update source field **(topic in two words)**
* :white_check_mark: (BSR) docs(github): add PR title valid format in template

</details>

<!-- Markdown tips:

To tick boxe replace [ ] with [x]

 -->

 <!--
TO DO: check with repo admin : [https://passculture.atlassian.net/browse/PC-<num>](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/configuring-autolinks-to-reference-external-resources)

JIRA-replace_with_ticket_number

[Notion-link](paste within parenthesis)
-->

lmontier-pass commented on jobs/ml_jobs/clusterisation/tools/clusterisation.py:
```diff
@@ -72,7 +73,11 @@ def mbkmeans_clusters(
+    km = MiniBatchKMeans(n_clusters=k, batch_size=mb, random_state=seed, verbose=1).fit(
+        X
+    )
+    logger.info(f"Random state for   MiniBatchKMeans fixed to seed={seed}")
```
"espace" en trop entre for et MiniBatchKMeans

lmontier-pass commented on jobs/ml_jobs/reduction/dimension_reduction.py:
```diff
@@ -132,6 +141,10 @@ def dimension_reduction(
         "default",
         help="String for the configuration plan to execute",
     ),
+    seed: int = typer.Option(
+        0,
+        help="Seed to fix random state of PCA space reduction",
+    ),
```
ne pas mettre la seed en paramètre du script. Plutôt la fixer et la logguer dans le fonction "bas" level qu'on puisse la retrouver

cibrahim-pass commented:
Changes:
* Added random seeds to PCA and MiniKMeans by generating a random seed with the secrets library. The secrest library is recommended by sonarqube. Secrets is native in python just like random module but itoffers a cryptographically strong. See [secrets module doc](https://docs.python.org/3/library/secrets.html#:~:text=In%20particular%2C%20secrets%20should%20be%20used%20in%20preference%20to%20the%20default%20pseudo%2Drandom%20number%20generator%20in%20the%20random%20module%2C%20which%20is%20designed%20for%20modelling%20and%20simulation%2C%20not%20security%20or%20cryptography.) 
* Removed he `-k` flag from dockerfiles so the curl [passes the security check](https://curl.se/docs/manpage.html#-k:~:text=pasv%2Dip.-,%2Dk%2C%20%2D%2Dinsecure,option%20makes%20curl%20skip%20the%20verification%20step%20and%20proceed%20without%20checking.,-When%20this%20option)

Tests done before PR:
* launched the algo-training-two-towers DAG on composer dev with this branch and [all tasks were successful.](https://3ecd2bbbcc0a45acb253336a797357bd-dot-europe-west1.composer.googleusercontent.com/dags/algo_training_two_towers/grid?tab=details&dag_run_id=manual__2024-12-02T10%3A20%3A03%2B00%3A00)
* make build to test the curl without the -k. Tested with netskope

lmontier-pass commented on jobs/ml_jobs/reduction/dimension_reduction.py:
```diff
@@ -18,31 +20,37 @@
 
 
 def reduce_transformation(
-    X, target_dimension, emb_col, method="PUMAP", max_dimension=32
+    X,
+    target_dimension,
+    emb_col,
+    method="PUMAP",
+    max_dimension=32,
 ):
```
ça serait peut être pas mal de documenter cette fonction (en tout cas la logique de la décision

* Si PCA => alors on réduit
* Si UMAP,PUMAP => On réduit via la PCA, puis par l'UMAP/PUMAP

lmontier-pass commented on jobs/ml_jobs/reduction/dimension_reduction.py:
```diff
@@ -18,31 +20,45 @@
+    method="PUMAP",
+    max_dimension=32,
 ):
+    """
+    Reduces X dimension according to method among ["PCA", "UMAP", "PUMAP"]
+    If chosen method is PCA:
+        - only PCA reduction is preformed
+    If chosen method is UMAP or PUMAP:
+        - First we reduce X with a PCA
+        - Then we reduce X AGAIN with UMAP or PUMAP
+    """
```
top. Dernier retour et après j'arrête... C'est top si tu peux ajouter à la docstring également les arguments et les return type. Par exemple tu peux regarder [ici](https://github.com/pass-culture/data-gcp/blob/3cec909b46863b09a89a5023b3e9ea0cdf855cb1/jobs/ml_jobs/artist_linkage/utils/clustering_utils.py)  comment c'est fait. 

Et si tu veux aller plus loin, on peut également typer les arguments de la fonction. 
```
reduce_transformation(
    X,
    target_dimension,
    emb_col,
    method="PUMAP",
    max_dimension=32,
):
```
->
```
reduce_transformation(
    X: np.ndarray,
    target_dimension: int,
    emb_col: str,
    method: str="PUMAP",
    max_dimension: int=32,
):
```

lmontier-pass commented on jobs/ml_jobs/reduction/dimension_reduction.py:
```diff
@@ -18,31 +20,58 @@
+        max_dimension (int): dimension threshold for X's second dimension to decide whether to perform a PCA before the UMAP/PUMAP or not
+
+    Returns:
+        X (np.ndarray): the reduced array
```
alors dans la typologie Returns:
tu ne mets pas les noms des variables

```
Returns:
        np.ndarray: the reduced array
```
